early-dev/app.R

```diff
@@ -187,181 +188,164 @@ fact_vars_asl <- names(Filter(isTRUE, sapply(ADSL, is.factor)))
+          AEREL2 = (AEREL == "Y" & ACTARM == "B: Placebo"),
+          ASTDT = as.Date(ASTDTM),
+          AENDT = as.Date(AENDTM)
+        ) %>%
+        teal.data::col_relabel(
+          RELFL = "Related AE",
+          CTC35FL = "Grade >=3 AE",
+          SERFL = "Serious AE",
+          RELSERFL = "Related Serious AE",
+          AEREL1 = "AE related to A: Drug X",
+          AEREL2 = "AE related to B: Placebo",
+          ASTDT = "Analysis Start Date",
+          AENDT = "Analysis End Date",
+          AESOC = "Primary System Organ Class",
+          AEBODSYS = "Body System or Organ Class",
+          AEHLT = "High Level Term",
+          AEDECOD = "Dictionary-Derived Term",
+          AETERM = "Reported Term for the Adverse Event",
+          AELLT = "Lowest Level Term"
+        )
+
+      ADCM <- synthetic_cdisc_data("latest")$adcm
+
+      # process ADCM
+      ADCM <- ADCM %>%
+        mutate(
+          ASTDT = as.Date(ASTDTM),
+          AENDT = as.Date(AENDTM)
+        ) %>%
+        teal.data::col_relabel(
+          ASTDT = "Analysis Start Date",
+          AENDT = "Analysis End Date"
+        )
+
+      ADEX <- synthetic_cdisc_data("latest")$adex
+
+      # process ADEX
+      ADEX <- ADEX %>%
+        mutate(
+          ASTDT = as.Date(ASTDTM),
+          AENDT = as.Date(AENDTM)
+        ) %>%
+        teal.data::col_relabel(
+          ASTDT = "Analysis Start Date",
+          AENDT = "Analysis End Date"
+        )
+
+      ADTR <- synthetic_cdisc_data("latest")$adtr
+
+      # process ADTR
+      adtr_labels <- teal.data::col_labels(ADTR, fill = FALSE)
+      ADTR <- ADTR %>%
+        mutate(
+          PCHG = ifelse(AVISIT == "BASELINE", 0, PCHG),
+          CHG = ifelse(AVISIT == "BASELINE", 0, CHG),
+          AVAL = ifelse(AVISIT == "BASELINE", BASE, AVAL),
+          AVALC = ifelse(AVISIT == "BASELINE", as.character(BASE), AVALC)
+        ) %>%
+        filter(AVISIT != "SCREENING")
+      teal.data::col_labels(ADTR) <- adtr_labels
+
+      ADTRWF <- ADTR %>%
+        filter(AVISIT != "BASELINE")
+      teal.data::col_labels(ADTRWF) <- teal.data::col_labels(ADTR, fill = FALSE)
+
+
+      # process ADRS
+      ADRSSWIM <- synthetic_cdisc_data("latest")$adrs %>%
+        filter(PARAMCD == "OVRINV") %>%
+        arrange(USUBJID)
+
+      ADRS <- synthetic_cdisc_data("latest")$adrs
+      adrs_labels <- teal.data::col_labels(ADRS, fill = FALSE)
+      ADRS <- ADRS %>%
+        filter(PARAMCD %in% c("BESRSPI", "INVET")) %>%
+        mutate(ADT = as.Date(ADTM)) %>%
+        droplevels()
+      teal.data::col_labels(ADRS) <- c(adrs_labels, "Analysis Date")
+
+      ADLB <- synthetic_cdisc_data("latest")$adlb
+
+      # process ADLB
+      ADLB <- ADLB %>%
+        mutate(
+          ADT = as.Date(ADTM),
+          LBSTRESN = as.numeric(LBSTRESC)
+        ) %>%
+        teal.data::col_relabel(
+          ADT = "Analysis Date",
+          LBSTRESN = "Numeric Result/Finding in Standard Units"
+        )
+    }),
```

Comment:
This is pretty bad that we have to repeat the code twice. Can we maybe came up with some good pattern how to not repeat ourselfs? Looking for some discussion here.

Reply:
If you do dynamic dots as suggested [here](https://github.com/insightsengineering/teal.data/pull/178#discussion_r1365290542) then you should be able to do the following:
```
code <- quote({
    ADSL <- synthetic_cdisc_data("latest")$adsl
    ADRS <- synthetic_cdisc_data("latest")$adrs
    ADLB <- synthetic_cdisc_data("latest")$adlb
    (...)
  })

data <- cdisc_data(
  !!!env_get_list(env = eval_to_env(code), nms = c("ADSL", "ADRS", "ADLB", "ADLBPCA")),
  code = code,
  join_keys = ...
)
```

Where
```
# this should be somewhere in the package
eval_to_env <- function(x, env = new.env()) {
  eval(x, envir = env)
  return(env)
}
```
We can even have a wrapper `eval_get_list <- function(x, nms){...}`

Reply:
I don't know what !!! means and frankly, I don't want to.

There is another option but it is a fresh idea and not refined yet.
Since `teal_data` is a `qenv`, perhaps the code could be run with `eval_code` only once?

Reply:
1. `!!!` is a splice operator -> [docs](https://rlang.r-lib.org/reference/splice-operator.html).
`foo(!!! list(a = 1, b = 2))` is equivalent to `foo(a = 1, b = 2)`. I really find it useful when dynamically parsing calls with ellipsis.

2. Can you elaborate more about multiple evaluation? Note that I just create `teal_data` (`qenv`) object and don't call `eval_code` method on it.
I expect no evaluation other than inside `eval_to_env` - in other words: I assume that `cdisc_data` (which calls `new_qenv`) will just assign parameters to returned object slots without any additional evaluation.

Reply:
Both
1. initializing `teal_data()` in the beginning of pre-processing and executing data-related-code using `eval_code(teal_data, expr)` or `within(teal_data, expr)`
2. separating data-code and app-elements (for example on two separate R files) and constructing `teal_data` based on `env` and `code`.

... are equivalent depending on the approach of app developer. Code can be executed in `teal_data` or in any other environment. There are still two evaluations as I wanted to make a PR which focus on recent changes (expecially to highlight join keys init).

I like the first one better as `teal_data` mimics a list so `data[["dataname"]]` can be used at any point in preprocessing.

Reply:
> I was thinking about
> 
> ```
> q <- new_qenv()
> q <- eval_code(q, <some_code>) 
> td <- <non_existent_function_that_makes_teal_data_out_of_qenv>(q)
> ```
> 
> but, like, @gogonzo mentioned it, we can just instantiate en empty `teal_data` and use `eval_code` on that. There is no use for the non-existent function as yet.

I was thinking about too but then I realised that oftentimes there are data-extract-spec objects in between that require access to the data objects created within qenv. While this is definitely accessible and doable I prefer not to require `teal.code` API familiarity (or at least minimise it as much as possible) (that's the option 1 from Dawid list). Hence I came up with the following:
```
code <- quote(...) # note this is a R base functionality
eval_to_env(code, global_env()) # side effects expected ## hopefully that's the only additional function required

arm_des <- data_extract_spec(...)
...

init(data = cdisc_data(adsl, ..., code = code), modules = foo_module(x = arm_des, ...), ...)
```

Reply:
So `eval_to_env` is just a wrapper for `eval` and works much like `eval_code` but without using `teal.code`. I would say this is needless duplication. `teal` does depend on `teal.data` and the `teal_data` class does inherit from `qenv`. Accessing data in a `qenv` is as easy as `q[["iris"]]` or `get_var(q, "iris")`.

Granted, we should update help pages in `teal.code` and that is coming.

Reply:
Currently `teal.code` is just _loaded_ and not _attached_ (which is good btw). That means that `teal.code` would have to be additionally attached with yet another `library()` call.
More importantly, IMHO `teal.code` should remain as an optional module development tool. The entry bar for app developer should be much lower. That's why I am not super convinced to make it an ~obligatory~ recommended package for the purpose of creating apps. The less we require the better. Regardless of its complexity.
Technically this would work out - I have no doubts. It's just a matter how simple is it from an app developer perspective.

Reply:
> Currently `teal.code` is just _loaded_ and not _attached_ (which is good btw). 

Good point. Perhaps `eval_code` should be re-exported from `teal.data`? `[[` and `within` methods become available on load, attaching is not required.

> More importantly, IMHO `teal.code` should remain as an optional module development tool.

I don't disagree. However, `teal_data` (the new one) is a child class. Should we expect the app dev to only treat `teal_data` as a container for data sets? Using the constructor requires the use of the `code` argument so some knowledge is indispensable.

Reply:
@pawelru @chlebowa I noticed a gap between our desirable code specification and what is now possible with `teal_data`. When initializing `teal_data` `@dataname`` slot is filled with data existing in the environment, so when we run `cdisc_data()` then `@datanames` will be empty.

```r
data <- cdisc_data()
data@datanames
#> character(0)
```

Evaluating any code in `teal_data` doesn't change `@datanames` and probably shouldn't because we want to know which objects in `@env` are real datasets and which are just help-objects. It is important to distinguish datasets from other objects for example to not create a filter-panel for data which is irrelevant.

```
data <- within(data, {
  ADSL <- synthetic_cdisc_data("latest")$adsl
  ADRS <- synthetic_cdisc_data("latest")$adrs
  ADLB <- synthetic_cdisc_data("latest")$adlb
})
```

So it means, one have to specify in `cdisc_data()` somehow what will be the `@datanames` im future environment.

Do you see my point?

This means that @pawelru initial proposition is easier to implement (to evaluate expression) and construct teal_data object from expression and data.

Reply:
Would this not also be an issue `teal_data` in general, not only with `cdisc_data`?
Perhaps we should open the ability to add data set names to the `@datanames` slot?
```
td <- teal_data()
td <- within(td, {
  i <- iris
  m <- mtcars
})
td <- add_datanames(td, c("i", "m"))
```

Reply:
Function `datanames()` to serve both set and get might be needed. We could implement following functions.



```
`datanames<-.teal_data` <- function(x, names) {
   if (all(names %in% ls(x@env))) stop()
   x@datanames <- names 
}

datanames.teal.data <- function(x) {
  if (length(x@datanames) == 0) {
    x@datanames
  } else {
    ls(x@env)
  }
}

```

By default `@datanames` can be empty and `datanames()` would return every binding name from @env.

Reply:
Something like that, details to be discussed in review. Shall I add these methods [here](https://github.com/insightsengineering/teal.data/pull/178)?